monitor/xfeemngr/gasprice/gasprice.go

```diff
@@ -0,0 +1,149 @@
+package gasprice
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/log"
+	"github.com/omni-network/omni/monitor/xfeemngr/streamer"
+
+	"github.com/ethereum/go-ethereum"
+)
+
+type Buffer struct {
+	mu sync.RWMutex
+
+	// map chainID to buffered gas price (not changed if outside threshold)
+	buffer map[uint64]uint64
+
+	// map chainID to provider
+	pricers map[uint64]ethereum.GasPricer
+
+	// options
+	opts *Opts
+}
+
+// NewBuffer creates a new gas price buffer.
+//
+// A gas price buffer maintains a buffered view of gas prices for multiple
+// chains. Buffered gas prices are not updated unless they are outside the
+// threshold percentage. Start steaming gas prices with Buffer.Stream(ctx).
+func NewBuffer(pricers map[uint64]ethereum.GasPricer, opts ...func(*Opts)) Buffer {
+	return Buffer{
+		buffer:  make(map[uint64]uint64),
+		pricers: pricers,
+		opts:    makeOpts(opts),
+	}
+}
+
+// GasPrice returns the buffered gas price for the given chainID.
+// If the price is not known, returns 0.
+func (b *Buffer) GasPrice(chainID uint64) uint64 {
+	p, _ := b.price(chainID)
+	return p
+}
+
+// Stream starts streaming gas prices for all providers into the buffer.
+func (b *Buffer) Stream(ctx context.Context) {
+	b.streamAll(ctx)
+}
+
+// streamAll starts streaming gas prices for all providers into the buffer.
+func (b *Buffer) streamAll(ctx context.Context) {
+	for chainID := range b.pricers {
+		b.streamOne(ctx, chainID)
+	}
+}
+
+// streamOne starts streaming gas prices for the given chainID into the buffer.
+func (b *Buffer) streamOne(ctx context.Context, chainID uint64) {
+	ctx = log.WithCtx(ctx, "chain", chainID)
+
+	callback := func(price uint64) {
+		if price == 0 {
+			return
+		}
+
+		// if price is buffed, and not outside threshold, return
+		buffed, ok := b.price(chainID)
+		if ok && !isOutsideThreshold(price, buffed, b.opts.thresholdPct) {
+			return
+		}
+
+		b.setPrice(chainID, price)
+	}
+
+	go subscribe(ctx, stream(ctx, b.pricer(chainID), b.opts.streamInterval), callback)
```

Comment:
I do not need a separate `stream` and `subcribe` that comm via a channel. We can just use the callback in the polling directly.

Channel was there becaes in previous design I was thinking of multiple subscribers

Reply:
Though I do kinda like the way they work together

Reply:
Yeah removing this now - can introduce if we need multiple subscribers